### PR TITLE
fix: jwt formats

### LINKED ISSUE #10: fix: jwt formats

### PR DESCRIPTION
**Pull request by chris13524**
# Description

The main concerns of mine are making sure the did:key|pkh stuff is correct and using right keys/formats. Also making sure the JWT claims fields are all correct. Also not sure if the ClientId stuff is redundant, seems odd.

I'm not really interested in code style feedback; I am just aiming for function in this PR as I inherited a lot of quickly-written code. Will cleanup as I continue working on the service. But I am happy to accept code style feedback on the auth.rs *Auth structures as I wrote those fresh.

Changes:

- Remove redundant PAT for actions/checkout
- Rename modules push_* -> notify_*
- notify_subscription response in JWT format
- notify_update request requires `act` = `notify_update`
- notify_update response in JWT format
- notify_delete request in JWT format
- notify_delete response in JWT format
- Fix lookup_table .replace_one() query. Because query was topic instead of _id it was querying for a row that never existed (topic doesn't exist in doc because it is renamed to _id by serde) it would attempt to insert the same data rather than upsert it.

Resolves #10 

## How Has This Been Tested?

Locally with integration tests

## Due Diligence

* [x] Breaking change
* [ ] Requires a documentation update
* [x] Requires a e2e/integration test update

### CODE REVIEW

**Review by TalhaAli00 — APPROVED** [review]

**Review by llbartekll — APPROVED** [review]

**Review by bkrem — APPROVED** [review]

File: src/auth.rs [hunk 1]
```diff
@@ -1,112 +1,253 @@
...
-    /// aud - did:key of an identity key. Enables to resolve associated Dapp
-    /// domain used.
+}
+
+pub trait GetSharedClaims {
+    fn get_shared_claims(&self) -> &SharedClaims;
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SubscriptionAuth {
```
Comment (Elyniss) [inline src/auth.rs:32]: nitpick: 
```suggestion
pub struct SubscriptionRequestAuth {
```

File: src/auth.rs [hunk 2]
```diff
@@ -1,112 +1,253 @@
...
+}
+
+impl GetSharedClaims for SubscriptionResponseAuth {
+    fn get_shared_claims(&self) -> &SharedClaims {
+        &self.shared_claims
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct UpdateAuth {
```
Comment (Elyniss) [inline src/auth.rs:78]: nitpick:
```suggestion
pub struct SubscriptionUpdateRequestAuth {
```

File: src/auth.rs [hunk 3]
```diff
@@ -1,112 +1,253 @@
...
+impl GetSharedClaims for UpdateAuth {
+    fn get_shared_claims(&self) -> &SharedClaims {
+        &self.shared_claims
+    }
+}
 
-        let header = base64::engine::general_purpose::STANDARD_NO_PAD.decode(header)?;
-        let header = serde_json::from_slice::<JwtHeader>(&header)?;
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct UpdateResponseAuth {
```
Comment (Elyniss) [inline src/auth.rs:102]: nitpick:
```suggestion
pub struct SubscriptionUpdateResponseAuth {
```

File: src/auth.rs [hunk 4]
```diff
@@ -1,112 +1,253 @@
...
+impl GetSharedClaims for UpdateResponseAuth {
+    fn get_shared_claims(&self) -> &SharedClaims {
+        &self.shared_claims
+    }
+}
 
-        let claims = base64::engine::general_purpose::STANDARD_NO_PAD.decode(claims)?;
-        let claims = serde_json::from_slice::<SubscriptionAuth>(&claims)?;
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct DeleteAuth {
```
Comment (Elyniss) [inline src/auth.rs:123]: nitpick:
```suggestion
pub struct SubscruptionDeleteRequestAuth {
```

File: src/auth.rs [hunk 5]
```diff
@@ -1,112 +1,253 @@
...
+    fn get_shared_claims(&self) -> &SharedClaims {
+        &self.shared_claims
+    }
+}
 
-        if claims.exp < Utc::now().timestamp().unsigned_abs() {
-            return Err(AuthError::Expired)?;
-        }
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct DeleteResponseAuth {
```
Comment (Elyniss) [inline src/auth.rs:144]: nitpick:
```suggestion
pub struct SubscriptionDeleteResponseAuth {
```

File: src/websocket_service/handlers/notify_subscribe.rs [hunk 6]
```diff
@@ -48,27 +58,63 @@ pub async fn handle(
         hex::encode(client_pubkey),
         project_data.signing_keypair.private_key,
     )?;
-    info!("[{request_id}] Response_sym_key: {}", &response_sym_key);
+    info!("[{request_id}] Response_sym_key: {}", &response_sym_key); // TODO don't log this
```
Comment (Elyniss) [inline src/websocket_service/handlers/notify_subscribe.rs:61]: Confused by the TODO, should be removed before deployment?

File: src/websocket_service/handlers/notify_subscribe.rs [hunk 7]
```diff
@@ -48,27 +58,63 @@ pub async fn handle(
...
+    let sub_auth = from_jwt::<SubscriptionAuth>(&msg.params.subscription_auth)?;
     let sub_auth_hash = sha256::digest(msg.params.subscription_auth);
 
-    // TODO verify `sub_auth.aud` matches `project_data.identity_keypair` (?)
+    if sub_auth.act != "notify_subscription" {
+        return Err(AuthError::InvalidAct)?;
+    }
+
+    // TODO verify `sub_auth.iss` matches SOMETHING (blockchain account?) because
+    // otherwise what's the purpose of having a signature
```
Comment (Elyniss) [inline src/websocket_service/handlers/notify_subscribe.rs:77]: iss is Identity Key. here you would call `verify_identity()`

**Review by Elyniss — APPROVED** [review]

File: src/websocket_service/handlers/notify_subscribe.rs (hunk 6 above)
Comment (chris13524) [reply src/websocket_service/handlers/notify_subscribe.rs:61]: It was already there, so I'm leaving it for now. Was planning to scan for more places where we log keys. But I could do this real quick in this PR.

File: src/websocket_service/handlers/notify_subscribe.rs (hunk 7 above)
Comment (chris13524) [reply src/websocket_service/handlers/notify_subscribe.rs:77]: Ahhhhh ok, that makes sense. I'll do that in a future PR as we discussed :)

File: src/auth.rs (hunk 5 above)
Comment (chris13524) [reply src/auth.rs]: d6d1aa4

File: src/auth.rs (hunk 4 above)
Comment (chris13524) [reply src/auth.rs]: d6d1aa4

File: src/websocket_service/handlers/notify_subscribe.rs (hunk 6 above)
Comment (chris13524) [reply src/websocket_service/handlers/notify_subscribe.rs]: 082ad28

**Pull request merged**